runners/spark/src/main/java/org/apache/beam/runners/spark/translation/MultiDoFnFunction.java

```diff
@@ -98,28 +110,108 @@ public MultiDoFnFunction(
 
     DoFnOutputManager outputManager = new DoFnOutputManager();
 
-    DoFnRunner<InputT, OutputT> doFnRunner =
+    SparkProcessContext.NoOpStepContext context;
+    InMemoryTimerInternals timerInternals = null;
+    // Now only implements the StatefulParDo in Batch mode.
+    if (stateful) {
+      Object key = null;
+      if (iter.hasNext()) {
+        WindowedValue<InputT> currentValue = iter.next();
+        key = ((KV) currentValue.getValue()).getKey();
+        iter = Iterators.concat(Iterators.singletonIterator(currentValue), iter);
+      }
+      final InMemoryStateInternals<?> stateInternals = InMemoryStateInternals.forKey(key);
+      timerInternals = new InMemoryTimerInternals();
+      final InMemoryTimerInternals finalTimerInternals = timerInternals;
+      context = new SparkProcessContext.NoOpStepContext(){
```

Comment:
This is not really a `NoOpStepContext` instance, but actually an implementation of `StepContext` (hence my comment about the type above).
```java
context = new StepContext(){
        @Override
        public StateInternals stateInternals() {
          return stateInternals;
        }

        @Override
        public TimerInternals timerInternals() {
          return timerInternals;
        }
      };
```
Might make more sense.